sigstore/_utils.py

```diff
@@ -172,3 +179,140 @@ def read_embedded(name: str, prefix: str) -> bytes:
     returning its contents as bytes.
     """
     return resources.files("sigstore._store").joinpath(prefix, name).read_bytes()  # type: ignore
+
+
+def cert_is_ca(cert: Certificate) -> bool:
+    """
+    Returns `True` if and only if the given `Certificate`
+    is a CA certificate.
+
+    This function doesn't indicate the trustworthiness of the given
+    `Certificate`, only whether it has the appropriate interior state.
+
+    This function is **not** naively invertible: users **must** use the
+    dedicated `cert_is_leaf` utility function to determine whether a particular
+    leaf upholds Sigstore's invariants.
+    """
+
+    # Only v3 certificates should appear in the context of Sigstore;
+    # earlier versions of X.509 lack extensions and have ambiguous CA
+    # behavior.
+    if cert.version != Version.v3:
+        raise InvalidCertError(f"invalid X.509 version: {cert.version}")
+
+    # Valid CA certificates must have *all* of the following set:
+    #
+    #  * `BasicKeyUsage.digitalSignature`
+    #  * `BasicKeyUsage.keyCertSign`
+    #  * `BasicConstraints.ca`
+    #
+    # Of those, non-CAs must have *only* `BasicKeyUsage.digitalSignature` set.
+    # Any other combination of states is inconsistent and invalid, meaning
+    # that we won't consider the certificate a valid non-CA leaf.
+
+    try:
+        basic_constraints = cert.extensions.get_extension_for_oid(
+            ExtensionOID.BASIC_CONSTRAINTS
+        )
+        ca = basic_constraints.value.ca  # type: ignore
+    except ExtensionNotFound:
+        # Only CA certificates should have BasicConstraints.
+        ca = False
+
+    digital_signature = False
+    key_cert_sign = False
+    try:
+        key_usage = cert.extensions.get_extension_for_oid(ExtensionOID.KEY_USAGE)
```

Comment:
`KeyUsage`'s criticality is actually a `SHOULD`, but this is perhaps somewhere that Sigstore as an ecosystem can be stronger.

cc @znewman01 @haydentherapper WDYT about a client side requirement that CA certs assert `KeyUsage.critical`? This shouldn't be a breaking change, since all existing CA certs should already have this extension marked correctly.